File: stp-project/frontend/src/types/health.ts
/**
 * Types mirroring backend/app/schemas/health.py.
 * Keep these two files in sync manually, or generate this file from the
 * OpenAPI schema (see README "Future Improvements") once the API is stable.
 */

export type HealthCategory =
  | "Healthy"
  | "Stable"
  | "Needs Attention"
  | "High Maintenance Risk";

export interface HealthProfile {
  component_id: string;
  repo_url: string | null;
  repo_resolved: boolean;

  scorecard_overall_score: number | null;
  contributors_count: number | null;
  is_archived: boolean;
  days_since_last_release: number | null;
  days_since_last_commit: number | null;

  maintenance_activity_score: number;
  release_cadence_score: number;
  community_resilience_score: number;
  security_hygiene_score: number;

  dhi_score: number;
  dhi_category: HealthCategory;
  explanation: string[];
}

export interface SbomHealthSummary {
  sbom_id: string;
  total_components: number;
  analyzed_components: number;
  category_breakdown: Record<string, number>;
  average_dhi_score: number;
  profiles: HealthProfile[];
}

File: stp-project/backend/app/schemas/health.py
"""
Pydantic schemas for the Dependency Health Intelligence API.

These are intentionally separate from the SQLAlchemy models: the API
contract should be free to evolve independently of the storage schema.
"""
import uuid

from pydantic import BaseModel, ConfigDict, Field

from app.models.health_profile import HealthCategory


class HealthProfileResponse(BaseModel):
    """Full health profile for a single component, returned to API clients."""

    model_config = ConfigDict(from_attributes=True)

    component_id: uuid.UUID
    repo_url: str | None
    repo_resolved: bool

    scorecard_overall_score: float | None
    contributors_count: int | None
    is_archived: bool
    days_since_last_release: int | None
    days_since_last_commit: int | None

    maintenance_activity_score: float
    release_cadence_score: float
    community_resilience_score: float
    security_hygiene_score: float

    dhi_score: float = Field(..., ge=0, le=100)
    dhi_category: HealthCategory
    explanation: list[str]


class SbomHealthSummary(BaseModel):
    """Aggregate health summary for every component in one SBOM."""

    sbom_id: uuid.UUID
    total_components: int
    analyzed_components: int
    category_breakdown: dict[str, int]
    average_dhi_score: float
    profiles: list[HealthProfileResponse]


class AnalyzeSbomRequest(BaseModel):
    """Request body to trigger a health analysis run for an SBOM."""

    force_refresh: bool = Field(
        default=False,
        description="If true, re-query external APIs even for components "
        "that already have a recent health profile.",
    )
